## 课程说明：

&emsp;&emsp;同学们好！欢迎来到《机器学习实战训练营》试学体验课！我是课程主讲老师，九天。        

- 本期公开课主题：LightGBM原理与实践

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202304301856951.png" alt="8ef6d1e3b063109eb052bb7021c01bd" style="zoom:33%;" />

&emsp;&emsp;LightGBM 是一种高效的 Gradient Boosting 算法，由 Microsoft Research Asia 团队开发，早期为Microsoft内部处理海量高维数据的专用算法，并于2017年由Guolin Ke, Qi Meng, Thomas Finley等人通过论文形式正式发布。如果说XGB为GBDT类算法在提升计算精度上做出了里程碑式的突破，那么LGBM则是在计算效率和内存优化上提出了开创性的解决方案，一举将GBDT类算法计算效率提高了近20倍、并且计算内存占用减少了80%，这也最终使得GBDT类算法、这一机器学习领域目前最高精度的预测类算法，能够真正应用于海量数据的建模预测。以下是官网给出的XGB、基于直方图优化的XGB和LGBM算法的在相同计算任务下计算时间的对比：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202304081540395.png" alt="0a747b1752ff4c7c9b368d7415b398c" style="zoom:40%;" />

而在内存占用方面，LGBM算法的优势也同样非常明显，以下是相同计算任务下不同算法的内存占用对比：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202304081540578.png" alt="a884b5b53ebd9665f526b82d74ca56b" style="zoom:40%;" />

而与此同时，LGBM能够保持和XGB几乎一样的预测精度，相同计算任务三种算法计算准确率如下：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202304081550947.png" alt="1680940216389" style="zoom:40%;" />

不难发现，预测精准而计算过程高效，这也是Light一词的核心精髓，并且经过这么多年的实践验证，可以说目前来看，LightGBM已然成为处理海量数据最高效、最通用的机器学习算法。

&emsp;&emsp;本期公开课，我们将围绕LightGBM算法的基本原理与实战技巧展开讨论，详细解读LightGBM算法原理，以及sklearn API的超参数优化及建模实战应用等。帮助大家快速了解LGBM算法精准高效的真正原因，并学习和掌握算法建模及优化技巧。

- 课程节选自《机器学习实战训练营》正式课程

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/1.jpg" alt="1" style="zoom:20%;" />

&emsp;&emsp;公开课内容节选自《机器学习实战训练营》正式课程，正式课程相比公开课，额外增加更具深度和LGBM数学原理推导、LGBM原生API使用和优化技巧，以及基于LGBM原生API的企业级实战案例，公开课与付费正课内容对比如下：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/05b61eb9b6792c10ce8c6c221f7d140.png" alt="05b61eb9b6792c10ce8c6c221f7d140" style="zoom:50%;" />

&emsp;&emsp;如果同学希望更进一步进行深度、系统的学习LightGBM及其他机器学习算法，欢迎大家报名《机器学习实战训练营（第六期）》付费正课，课程为130+小时完整体系大课，完整涵盖经典机器学习、集成学习、特征工程、模型融合、特征混合增强和企业级之战案例，总共6大模块，零基础直达机器学习算法岗中高级岗位要求，六年教学经验沉淀，千名学员口碑见证，超值体系大课等你来学！

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230215185442588.png" alt="image-20230215185442588" style="zoom:40%;" />

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230215185629296.png" alt="image-20230215185629296" style="zoom:42.5%;" />

&emsp;&emsp;除了《机器学习实战训练营》课程外，我们还开设了《数据分析就业班》和《深度学习实战课》，在此查看详细课程介绍和课程大纲：https://appZe9inzwc2314.h5.xiaoeknow.com：

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230214180459436.png" alt="image-20230214180459436" style="zoom:50%;" />

- **<font color='red'>第六期训练营课程尾单八折特惠进行时！**

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/2.jpg" alt="2" style="zoom:30%;" />

&emsp;&emsp;<font size=3><center>扫码添加客服“小可爱”，或者手动添加客服微信：littlecat_1205       
    >>>  <font color='red'>**回复“优惠”**</font>抢<font color='red'>**直播间限定折上折**        
    **>>>  限量赠送授课老师亲自答疑服务！**

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/926380675d5976b65f5784be6db5f35.png" alt="926380675d5976b65f5784be6db5f35" style="zoom:50%;" />

---

# <center> LightGBM原理与实践

# <center>Day 1.LightGBM基本原理

- LightGBM原理简介

&emsp;&emsp;LightGBM（Light Gradient Boosting Machine，以下简称LGBM）是一个基于梯度提升决策树（Gradient Boosted Decision Trees，GBDT）的高效、可扩展的机器学习算法，作为GBDT框架的算法的一员，并且作为XGB算法的后来者，LGBM非常好综合了包括XGB在内的此前GBDT算法框架内各算法的一系列优势，并在此基础上做了一系列更进一步的优化。LGBM算法提出的核心目的是为了解决GBDT算法框架在处理海量数据时计算效率低下的问题，而从实践效果来看，LGBM也确实做到了这点——LGBM以牺牲极小的计算精度为代价，将GBDT的计算效率提升了近20倍！这也最终使得LGBM算法是第一个真正意义上能处理海量数据的GBDT框架算法。并且，尽管计算精度存在“选择性的牺牲”，但LGBM的实际建模效果也能达到几乎和XGB同等水平，而且由于LGBM“选择性的牺牲精度”从另一个角度来看其实就是抑制模型过拟合，因此在很多场景下，LGBM的算法效果甚至会好于XGB。种种实践证明，LGBM是一个拥有超高计算效率的同时、又能够保持超高精度的算法，是目前机器学习领域当之无愧的顶级算法之一。

&emsp;&emsp;而LGBM是如何做效率和精度“两手抓”的呢？简而言之就是LGBM充分借鉴了XGB提出的一系列提升精度的优化策略，同时在此基础之上进一步提出了一系列的数据压缩和决策树建模流程的优化策略。尽管在算法的数学原理层面LGBM并没有翻越XGB创建的理论高峰，但其提出的一系列优化策略也同样是极具开创性的，其中数据压缩方法能够让实际训练的数据量在大幅压缩的同时仍然保持较为完整的信息，而决策树建模流程方面的优化，则是在XGB提出的直方图优化算法基础上进行了大幅优化，不仅能够加速决策树建模速度，同时也能非常好的处理经过压缩后的数据，从而最终大幅提升每棵树的训练效率（甚至在LGBM提出的一段时间后，新版XGB也采用了LGBM类似的直方图算法来加速建模效率）。并且最重要的是，有理论能够证明，哪怕LGBM实际建模是基于压缩后的数据进行训练，但其预测精度受到的影响也是微乎其微。

&emsp;&emsp;当然，除了算法原理层面的优化方法外，LGBM还提出了非常多针对于实际计算过程的优化，例如Voting Parallel（投票特征并行）方法、特征多线程并行处理方法、GPU加速方法和分布式计算等，这些方法进一步提升了LGBM实际建模效率，并且一定程度拓宽了算法的使用场景。并且需要注意的是，所谓的计算效率优化，不仅体现在计算时间的大幅缩短，同时得益于LGBM所提出的一系列数据压缩技术，使得实际建模时数据内存占用也大幅减少。

&emsp;&emsp;总的来说，LGBM算法可以看成是迭代过程几乎全盘借鉴XGB、而在数据压缩方法和决策树训练方法上有大量创新的算法，因此在原理相关内容我们将分为两部分进行讲解，第一部分我们重点介绍LGBM创新性提出的一系列方法，第二部分再来探讨LGBM损失函数求解流程。考虑到LGBM的推导流程和XGB几乎完全一样，原理部分的讲解的重点将会是LGBM创新性提出的一系列数据压缩和优化策略。

- LightGBM的数据压缩策略

&emsp;&emsp;LightGBM建模过程总共会进行三方面的数据压缩，根据实际建模顺序，会现在全样本上连续变量分箱（连续变量离散化），然后同时带入离散特征和离散后的连续变量进行离散特征捆绑（合并）降维，最终在每次构建一颗树之前进行样本下采样。其中连续变量的分箱就是非常简单的等宽分箱，并且具体箱体的数量可以通过超参数进行人工调节；而离散特征的降维，则是采用了一种所谓的互斥特征捆绑（Exclusive Feature Bundling, EFB）算法，该算法也是由LGBM首次提出，该方法的灵感来源于独热编码的逆向过程，通过把互斥的特征捆绑到一起来实现降维，这种方法能够很好的克服传统降维方法带来的信息大量损耗的问题，并且需要注意的是，输入EFB进行降维的特征，即包括原始离散特征，也包括第一阶段连续变量离散化之后的特征；在这一系列数据压缩之后，LGBM在每次迭代（也就是每次训练一颗决策树模型）的时候，还会围绕训练数据集进行下采样，此时的下采样不是简单的随机抽样，而是一种名为基于梯度的单边采样（Gradient-based One-Side Sampling, GOSS）的方法，和EFB类似，这种方法能够大幅压缩数据，但同时又不会导致信息的大量损失。不难发现，最终输入到每颗决策树进行训练的数据，实际上是经过大幅压缩后的数据，这也是LGBM计算高效的根本原因之一。

- LightGBM决策树建模优化方法

&emsp;&emsp;而进入到具体的决策树训练环节，总的来说LGBM采用的决策树建模优化方法有两个，其一是直方图优化算法，这种方法本质上是通过直方图的形式更加高效简洁的表示每次分裂前后数据节点的核心信息，并且父节点和子节点也可以通过直方图减法计算直接算得，从而加速数据集分裂的计算过程:

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303221537628.png" alt="981861843f708f9efb5f74829c336b3" style="zoom: 40%;" />

其二则是leaf wise tree growth的叶子节点优先的决策树生长策略，这其实是一种树生长的模式，对于其他大多数决策树算法和集成算法来说，树都是一次生长一层，也就是所谓的Level-wise tree growth（深度优先的生长策略），生长过程如下。

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303221541288.png" alt="b4030247c4f8acaee3a8337bb6e7bb2" style="zoom:50%;" />

而LGBM则允许决策树生长过程优先按照节点进行分裂，即允许决策树“有偏”的生长，也就是所谓的leaf wise tree growth的叶子节点优先的决策树生长策略，具体生长过程如下：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303221543679.png" alt="33f0d4c90d591b2577698a730d24dfc" style="zoom:50%;" />

根据LGBM论文的论述，但从Level-wise tree growth远离层面，这种方法其实是有利有弊，其优势在于能够大幅提升每颗树的收敛速度，从总体来看相当于是提升了每次迭代效率；而问题则在于会每棵树的计算过程会变得更加复杂，并且存在一定的过拟合风险。不过对于LGBM来说，这些问题都能够被很好的克服，比如计算过程复杂的问题可以通过数据压缩来对冲，而过拟合风险则可以通过限制最大树深度来解决，因此总的来看Level-wise tree growth就是最适合LGBM的决策树生长策略。

&emsp;&emsp;接下来，我们就这些技术逐一来进行介绍，并借助一个精心设计的手动实现的例子，来串联起LGBM在进行Boosting迭代前的全部数据计算流程，借此深化大家对本部分内容的理解。

### 1.连续变量分箱

- 等宽分箱基本概念回顾

&emsp;&emsp;首先是连续变量分箱。LGBM采用的连续变量分箱方法就是简单的等宽分箱，和我们在特征工程部份介绍的等宽分箱方法无异：首先计算连续变量的取值范围，然后人工输入的max_bin超参数，进行数量为max_bin等宽度的区间划分，并把连续变量的值划归到一个个箱体内部。例如某连续变量取值范围为[0, 10]，max_bin=2，则两个等宽的区间划分为bin0=[0, 5)和bin1=[5, 10]，并且如果某连续变量取值为1，则经过分箱后会被标记为bin0（或者0），如果某各连续变量取值为10，则分箱后会被标记为bin1（或者1）。至此，就将连续变量转化为了离散变量。具体手动实现过程和sklearn实现过程可以回顾特征工程Part 2中的内容介绍。

> 这里需要注意，XGB也会对连续变量进行分箱，但XGB的分箱是分位数分箱，而不是等宽分箱。

- 手动示例

&emsp;&emsp;接下来，我们通过一个手动实现的例子来说明这一过程，需要注意的是，这个手动创建数据集将贯穿本节的各部分内容。数据集基本情况如下：

In [45]:
np.random.seed(11)

x1 = np.array([1.2, 2.9, 2.6, 3.3, 2.0, 2.5, 1.4, 2.1, 1.7, 3.0])
x2 = np.array([4.7, 5.5, 3.9, 6.2, 3.5, 4.5, 5.1, 2.7, 4.1, 3.8])
x3 = np.random.randint(0, 2, 10)
x4 = np.random.randint(0, 2, 10)
y = np.array([1, 0, 1, 0, 1, 1, 0, 0, 1, 1])
data = pd.DataFrame({'x1':x1, 'x2':x2, 'x3':x3, 'x4':x4, 'y':y})
data

x1   x2  x3  x4  y
0  1.2  4.7   1   0  1
1  2.9  5.5   1   0  0
2  2.6  3.9   0   1  1
3  3.3  6.2   1   0  0
4  2.0  3.5   1   0  1
5  2.5  4.5   1   1  1
6  1.4  5.1   1   0  0
7  2.1  2.7   0   1  0
8  1.7  4.1   1   0  1
9  3.0  3.8   1   1  1

数据集总共包含10条数据，其中x1和x2是连续特征，x3和x4是离散特征，y是标签，该数据集是各二分类数据集。

&emsp;&emsp;接下来，我们对其中的连续变量进行分箱，这里我们设置max_bin=2，即进行两个箱体的等宽分箱。具体实现过程及分箱结果如下：

In [46]:
from sklearn.preprocessing import KBinsDiscretizer

# 将 x1 和 x2 分别进行等宽分箱，分成 2 个箱子
n_bins = 2
strategy = 'uniform'

kbins_x1 = KBinsDiscretizer(n_bins=n_bins, encode='ordinal', strategy=strategy)
kbins_x2 = KBinsDiscretizer(n_bins=n_bins, encode='ordinal', strategy=strategy)

x1_binned = kbins_x1.fit_transform(data['x1'].values.reshape(-1, 1))
x2_binned = kbins_x2.fit_transform(data['x2'].values.reshape(-1, 1))

# 将分箱后的结果保存到原始数据集中
data['x1_binned'] = x1_binned
data['x2_binned'] = x2_binned

print(data)

    x1   x2  x3  x4  y  x1_binned  x2_binned
0  1.2  4.7   1   0  1        0.0        1.0
1  2.9  5.5   1   0  0        1.0        1.0
2  2.6  3.9   0   1  1        1.0        0.0
3  3.3  6.2   1   0  0        1.0        1.0
4  2.0  3.5   1   0  1        0.0        0.0
5  2.5  4.5   1   1  1        1.0        1.0
6  1.4  5.1   1   0  0        0.0        1.0
7  2.1  2.7   0   1  0        0.0        0.0
8  1.7  4.1   1   0  1        0.0        0.0
9  3.0  3.8   1   1  1        1.0        0.0


至此，我们就完成了LGBM的第一阶段数据处理——连续变量的分箱处理。

### 2.互斥特征捆绑（Exclusive Feature Bundling，EFB）

&emsp;&emsp;接下来则是围绕这些离散特征进行降维。LGBM采用了一种名为互斥特征捆绑（Exclusive Feature Bundling，EFB）的降维方法，这种方法在LGBM论文LightGBM: A Highly Efficient Gradient Boosting Decision Tree (2017)中首次提出，不同于第一阶段的简单的等宽分箱，EFB实际计算过程非常复杂，我们这里从EFB方法提出背景、计算原理和手动示例三个方面对其进行介绍。

#### 2.1 EFB算法简介与基本流程

- EFB算法提出背景

&emsp;&emsp;根据LightGBM: A Highly Efficient Gradient Boosting Decision Tree (2017)论文描述，原始的GBDT在进行每颗树的训练时，需要带入全部数据来进行信息增益的计算，从而寻找到决策树生长过程中的最佳切分点，这个过程也就是所谓的扫描全部数据来决定切分点的过程。这个过程尽管非常精准，但计算复杂度非常高（直接和特征数量及样本数量成正比），在进行海量数据建模训练的时候会耗费大量的算力和时间。因此，为了能够更好的应对海量数据的模型训练，样本采样和特征降维是非常必要的。但传统的方法在这方面往往效果不佳，例如简单的欠采样（样本随机抽样）可能会造成模型训练过程非常不稳定，而PCA降维则只适用于处理冗余特征，当每个特征都具有相当信息体量时强行进行降维则会导致信息大量丢失。为了解决这个问题，LGBM开创性的提出了基于梯度的单边采样方法（GOSS）进行样本数量的压缩，提出了互斥特征捆绑方法（EFB）来进行特征压缩。不同于以往的方法，GOSS和EFB能够非常好的兼顾预测精度与计算效率。此外，对连续变量进行离散化也是非常有效的数据压缩的手段，LGBM在XGB提出的直方图优化的基础上，进一步提出了一种改进策略，和GOSS和EFB类似，这种LGBM直方图优化方法同样能够在大幅提高计算效率的同时保证预测精度。

- 简化后的EFB计算流程

&emsp;&emsp;而具体到EFB降维算法，其实是受到独热编码启发，设计的类似于独热编码逆向过程的一种算法。例如一组数据情况如下，独热编码是从左往右的计算过程，把一列展开为多列，而EFB则是从右往左进行计算，将多列压缩为一列：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303222002280.png" alt="0be767e9953d7db2a66752f524e6ba4" style="zoom:50%;" />

> 具体独热编码的相关内容，可回顾特征工程Part 2数据重编码部分内容。

&emsp;&emsp;那既然是独热编码的逆向计算，我们就需要首先讨论为什么LGBM不需要独热编码。我们知道，独热编码本质上是对离散特征更加细粒度的信息呈现，在某些场景下能够提升模型效果。当然更重要的是独热编码能够非常好的用于表示离散变量，对于大多数无法区分连续变量和离散变量的机器学习算法来说，通过独热编码重编码的数据将能够非常方便进行例如离散变量之间的距离计算等操作。但是这些独热编码的优势对于LGBM来说并不存在。首先LGBM带入模型计算的全部变量都是离散变量（连续变量也会被离散化），其次独热编码带来的更细粒度的信息呈现也不会进一步提升模型效果（对于大多数集成学习算法来说都是如此），当然更重要的是，LGBM的算法设计就是为了处理海量高维数据，独热编码只会进一步造成维度灾难。因此，LGBM不仅不需要进行独热编码，还需要进行独热编码的逆操作来进行特征降维。当然，我们这里只是借用独热编码的计算过程帮大家理解EFB的降维过程，在实际计算过程中，EFB的降维的目标并不是把独热编码之后的特征再转换回来，而是找到那些原始状态下就存在类似上图中x1和x2这种关系的特征，来将其合成为一个特征。这里我们注意观察，上图中x1和x2两个特征存在一种这样的关系——任何一条样本都不会同时在x1和x2上取值非0，在EFB原理的定义中，这种关系的特征又被称作互斥特征，而互斥特征其实是可以合成一个特征的，比如上图中的x，这个合成的过程并不会有任何的信息损失，而合成的过程又被称作特征捆绑。这也就是所谓的互斥特征捆绑算法。

&emsp;&emsp;我们这里再看一个互斥特征捆绑的例子，比如如下x3和x4，也是互斥的，此时我们可以将x3和x4捆绑为一个新的x_b1特征，新特征中可以用0、1、2来表示x3和x4的不同组合，从而在不损失信息的情况下，进行了降维。

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303222003609.png" alt="6a6d2ee962754f94b7232cb069835ce" style="zoom:50%;" />

&emsp;&emsp;当然，这只是一个简化后的示例，真实的EFB特征降维情况会非常复杂，并不是简单的将多个离散变量的不同取值组合进行重新赋值，这个例子只是用于帮大家建立对EFB的感性的认识，接下来我们就围绕原论文中提出的EFB算法来进行更加严谨的算法流程介绍。

#### 2.2 EFB算法基本原理

- 放宽互斥的条件：冲突比例（conflict_rate）概念介绍

&emsp;&emsp;真实数据的EFB计算过程会非常复杂，首先是关于“互斥”关系的定义，EFB并不是只压缩完全互斥的特征，而是非常灵活的定义了一个冲突比例（又称非互斥比例），这个比例用于表示两个特征中冲突（即非互斥、同时取非零值）的取值占比，来衡量两个特征互斥程度。当然，冲突比例越大说明互斥程度越低。例如对于如下数据集，总共包含四条数据，其中只有第四条数据是同时取得了非零值，因此只有一条数据是冲突的，其他数据都是互斥的，因此冲突比例为1/4=0.25：

|特征1|特征2|      
|:--:|:--:|
|0|1|
|1|0|
|0|1|
|1|1|

而如果数据集如下，此时两个特征非同时为0的（非全零样本）数量为3，则冲突比例为1/3:

|样本情况|特征1|特征2|      
|:--:|:--:|:--:|
|非全零样本|0|1|
|非全零样本|1|0|
|全零样本|0|0|
|非全零样本|1|1|

同时，LGBM提供了一个名为max_conflict_rate的超参数，用于表示最大冲突比例，当两个特征的冲突比例小于我们设置的最大冲突比例（max_conflict_rate）时，我们就认为这两个特征并不冲突，而是互斥的，是可以进行捆绑的。例如假设我们设置max_conflict_rate=0.3，则上述两个特征可以进行捆绑，而如果我们设置max_conflict_rate=0.2，则上面两个特征超过了我们认为冲突的阈值，因此这两个特征是冲突的，而不是互斥的，是不能进行进一步捆绑的。很明显，max_conflict_rate设置的越小，对互斥的要求就越高，特征就越不容易被捆绑，模型计算量就越大、精度也越高，而如果max_conflict_rate设置的很大，则更多的特征会被捆绑，模型计算速度会更快，但精度也会降低。

- 借助图着色(Graph Coloring Problem)问题来解决特征捆绑流程问题

&emsp;&emsp;通过最大冲突比例的概念引入，相当于是放宽的互斥的条件，或者说给是否互斥添加了一个可以量化计算的阈值。而真正开始进行特征捆绑的时候，面对海量特征，LGBM是如何进行EFB计算的呢？实际上LGBM会将特征捆绑问题视作（或者说是转化为）一个图着色的问题(Graph Coloring Problem)。图着色问题一种经典的组合优化问题，其问题描述为：给定一个无向图，如何用尽量少的颜色对图中的每个顶点进行着色，使得相邻的顶点颜色不同。这里的“颜色”可以是任意一种符号或编号，只要保证相邻的顶点颜色不同即可。在EFB计算过程中，会将不同特征视作图上的一个个点，若特征之间存在冲突，则用一条无向边进行连接，边的权重就是冲突比例，而如果两个特征互斥，则彼此没有边进行相连。而在将特征及其冲突情况用图进行展示后，即可进一步进行图着色——即在相邻的点颜色不同的前提条件下，用尽可能少的颜色对图上的点进行着色，既然相互冲突的特征都有边进行相连，那么相同颜色的点其实就是互斥的特征，接下来我们仅需把相同颜色的特征进行合并即可。

&emsp;&emsp;当然，这个过程听起来较为抽象，我们这里还是以上面的data数据集为例，来展示一个借助图着色来进行特征捆绑的完整过程。

#### 2.3 EFB算法计算流程

- 计算冲突比例矩阵

&emsp;&emsp;经过连续变量离散化，现在我们的data数据就已经变成了四个离散特征的数据集，四个离散特征分别为：'x1_binned'、'x2_binned'、'x3'、'x4'：

In [47]:
data

x1   x2  x3  x4  y  x1_binned  x2_binned
0  1.2  4.7   1   0  1        0.0        1.0
1  2.9  5.5   1   0  0        1.0        1.0
2  2.6  3.9   0   1  1        1.0        0.0
3  3.3  6.2   1   0  0        1.0        1.0
4  2.0  3.5   1   0  1        0.0        0.0
5  2.5  4.5   1   1  1        1.0        1.0
6  1.4  5.1   1   0  0        0.0        1.0
7  2.1  2.7   0   1  0        0.0        0.0
8  1.7  4.1   1   0  1        0.0        0.0
9  3.0  3.8   1   1  1        1.0        0.0

In [48]:
data[['x1_binned', 'x2_binned', 'x3', 'x4']]

x1_binned  x2_binned  x3  x4
0        0.0        1.0   1   0
1        1.0        1.0   1   0
2        1.0        0.0   0   1
3        1.0        1.0   1   0
4        0.0        0.0   1   0
5        1.0        1.0   1   1
6        0.0        1.0   1   0
7        0.0        0.0   0   1
8        0.0        0.0   1   0
9        1.0        0.0   1   1

然后我们需要计算这四个特征彼此之间的冲突比例，我们可以通过定义如下函数来完成计算。该函数定义了多个特征彼此之间冲突比例矩阵的计算过程，这里的冲突比例矩阵就类似于相关系数矩阵，用于表示多个特征彼此之间冲突比例：

In [49]:
def conflict_ratio_matrix(data):
    """
    冲突比例计算函数
    
    :param data: 计算冲突比例的dataframe
    :return:冲突比例矩阵
    """
    if isinstance(data, pd.DataFrame):
        data = data.values

    num_features = data.shape[1]
    # 创建空特征比例矩阵
    conflict_matrix = np.zeros((num_features, num_features))

    # 两层循环挑选两个特征
    for i in range(num_features):
        for j in range(i+1, num_features):
            # 计算特征冲突特征总数
            conflict_count = np.sum((data[:, i] != 0) & (data[:, j] != 0))
            # 计算不为0的特征总数
            total_count = np.sum((data[:, i] != 0) | (data[:, j] != 0))
            # 计算冲突比例
            conflict_ratio = conflict_count / total_count
            conflict_matrix[i, j] = conflict_ratio
            conflict_matrix[j, i] = conflict_ratio

    return conflict_matrix

然后尝试带入四个离散变量，进行冲突比例的计算：

In [50]:
conflict_ratio_matrix(data[['x1_binned', 'x2_binned', 'x3', 'x4']])

array([[0.        , 0.42857143, 0.44444444, 0.5       ],
       [0.42857143, 0.        , 0.625     , 0.125     ],
       [0.44444444, 0.625     , 0.        , 0.2       ],
       [0.5       , 0.125     , 0.2       , 0.        ]])

其中，矩阵中的第(i, j)个元素代表第i个特征和第j个特征的冲突比例，例如(1, 2)=0.42857表示第一个特征和第二个特征冲突比例为0.42857。能够看出，这四个特征中并不存在没有冲突（即互斥）的特征，只能看到2、4和3、4特征冲突比例较小：

In [51]:
data[['x1_binned', 'x2_binned', 'x3', 'x4']]

x1_binned  x2_binned  x3  x4
0        0.0        1.0   1   0
1        1.0        1.0   1   0
2        1.0        0.0   0   1
3        1.0        1.0   1   0
4        0.0        0.0   1   0
5        1.0        1.0   1   1
6        0.0        1.0   1   0
7        0.0        0.0   0   1
8        0.0        0.0   1   0
9        1.0        0.0   1   1

- 图展示与着色

&emsp;&emsp;然后我们将冲突比例矩阵转化为如下所示的图，即不同的点代表着不同的特征，而如果两个特征存在冲突，则两个点之间构建一条无向边，边的权重就是冲突比例：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303231529895.png" alt="1679556539065" style="zoom:50%;" />

当然，关于是否互斥，其实可以通过max_conflict_rate进行调节，我们假设max_conflict_rate=0.3，则上图中x3和x4、x2_binned和x4的冲突比例小于0.3，所以这两组特征是互斥的，我们可以将连接这两组特征的边删除，删除后的图如下：

> 这里需要注意，这里我们设置的max_conflict_rate=0.3只是用于当前例子展示所用。真实情况下max_conflict_rate的取值建议设置为0.1或者更小的数，以确保模型精度。相关超参数取值推荐会在本章课程的最后进行讨论。

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303231538803.png" alt="1679557086174" style="zoom:50%;" />

在完成图转化之后，接下来我们将特征捆绑问题视作一个图着色的问题，即需要用最少的颜色对图上的四个点进行着色，并要求相邻的点（彼此有线段连接的点）颜色不同。

&emsp;&emsp;具体着色的流程是会先从度（边的个数，也被成为degree）更大的点进行着色，例如上图中的四个点的degree如图所示，很明显x1_binned的度最大，然后是x2_binned和x3，这里我们先将x1_binned着色为红色（颜色可以随机选取），然后由于x3和x2_binned彼此相连，并且都和x1_binned相连，因此x3和x2_binned颜色不能相同，且不能和x1_binned相同，因此这里分别给x3和x2_binned着色为黄色和绿色，最后是x4，由于x4和x1_binned相连，所以x4不能用红色，而x4和x3、x2_binned没有相连，并且为了图上出现的颜色尽可能少，所以x4可以用绿色或者黄色，考虑到x4和x2_binned冲突比例较小，互斥程度较大，因此可以给x4使用绿色，最后着色结果如上图所示。

- 特征捆绑过程

&emsp;&emsp;最后，我们把相同着色的点（特征）进行捆绑。捆绑过程并不复杂，核心是需要对特征取值进行合理转化。而具体的转化过程中，LGBM会根据主特征的最大取值设置一个offset值，然后对合并进来特征的非零值加上这个offset值，然后再让这两个特征取值相加。例如data数据集中，我们将x4并入x2_binned中，则offset=1

In [52]:
data['x2_binned']

0    1.0
1    1.0
2    0.0
3    1.0
4    0.0
5    1.0
6    1.0
7    0.0
8    0.0
9    0.0
Name: x2_binned, dtype: float64

然后对x4的非零值+offset，并构成新的特征'x2_binned&x4'具体计算过程如下：

In [53]:
arr = np.array(data['x4'])
arr[arr != 0] += 1
arr

array([0, 0, 2, 0, 0, 2, 0, 2, 0, 2])

In [54]:
data['x2_binned&x4'] = arr + data['x2_binned']
data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4
0  1.2  4.7   1   0  1        0.0        1.0           1.0
1  2.9  5.5   1   0  0        1.0        1.0           1.0
2  2.6  3.9   0   1  1        1.0        0.0           2.0
3  3.3  6.2   1   0  0        1.0        1.0           1.0
4  2.0  3.5   1   0  1        0.0        0.0           0.0
5  2.5  4.5   1   1  1        1.0        1.0           3.0
6  1.4  5.1   1   0  0        0.0        1.0           1.0
7  2.1  2.7   0   1  0        0.0        0.0           2.0
8  1.7  4.1   1   0  1        0.0        0.0           0.0
9  3.0  3.8   1   1  1        1.0        0.0           2.0

至此，我们就在data这个简化的数据集上完成了EFB特征捆绑过程，经过连续变量分箱和特征捆绑，实际上接下来带入进行模型训练的特征就只有x1_binned、x2_binned&x4和x3这三个特征：

In [55]:
data[['x1_binned', 'x2_binned&x4', 'x3']]

x1_binned  x2_binned&x4  x3
0        0.0           1.0   1
1        1.0           1.0   1
2        1.0           2.0   0
3        1.0           1.0   1
4        0.0           0.0   1
5        1.0           3.0   1
6        0.0           1.0   1
7        0.0           2.0   0
8        0.0           0.0   1
9        1.0           2.0   1

### 3.基于梯度的单边采样（Gradient-based One-Side Sampling，GOSS）

#### 3.1 GOSS计算基本原理

&emsp;&emsp;首先来看基于梯度的单边采样（Gradient-based One-Side Sampling）方法，也就是所谓的GOSS抽样方法。不同于简单随机抽样，GOSS是一种非常特殊的基于梯度分布的抽样方法。我们知道在执行优化算法的过程中，每个样本都有一个对应的梯度计算结果，如果某条样本梯度绝对值较小，则说明这条样本已经被正确的分类或者预测结果和真实结果非常接近，在后续的参数更新过程中，这些梯度绝对值较小的样本对参数的改进贡献较小，因此每次迭代计算时再把这些小梯度的样本再算一遍梯度，会一定程度造成资源浪费。而反观那些梯度绝对值较大的样本，这些样本具有更高的误差，因此对模型的训练有更大的贡献。因此GOSS的思路是将全部样本按照梯度绝对值大小进行降序排序，然后抽取梯度绝对值最大的前$a\%$的样本，然后把其他样本都视作小梯度样本，并从这些小梯度样本中随机抽取$b\%$个样本，而这些大梯度样本和随机抽取的小梯度样本，就构成了接下来模型训练的数据集。而只针对小梯度样本（一边）进行抽样、保留（另一边）全部大梯度样本，也就是单边采样一词的由来。

&emsp;&emsp;而在具体执行GOSS的时候有以下几点需要注意：         

- 1.GOSS计算过程是根据梯度的绝对值大小进行样本划分和抽样，并不是样本梯度的真实值；

- 2.GOSS中样本选取比例，也就是梯度绝对值最大的前$a\%$和小样本中随机抽样的$b\%$，实际上都是超参数，可以在建模过程中灵活调节。这里的$a\%$可以换成更专业的超参数名称：top_rate，而小样本抽取的$b\%$更专业的名称则是other_rate；     

- 3.我们知道样本梯度是基于预测结果计算而来的（具体来说是损失函数的一阶导数），而在第一棵树构建之前我们就需要进行GOSS采样，此时还没有模型预测结果，梯度计算依据的是LGBM的初始预测值，和其他集成学习类似，LGBM的初始预测值也是根据损失函数的不同类型计算得到的结果；

- 4.由于每次迭代都会更新模型参数，因此每次建树之前都会重新进行抽样，而除非人为控制迭代过程（例如使用一种非常特殊的Booster API，后面会详细介绍），否则一般来说top_rate和other_rate设置好了就不会再发生变化；     

- 5.关于top_rate和other_rate的数值设置，一般来说top_rate越大other_rate越小，则模型过拟合风险就越大，反之则模型学习能力会被限制，而如果这两个参数同时较大，则会增加模型训练复杂度，增加模型训练时间。关于这组参数，并没有一个普遍适用的取值，还是需要根据实际情况进行超参数优化。

- 6.尽管带入训练的数据是GOSS抽样后的数据，但在后续决策树生长的过程中，小梯度样本的梯度（和损失函数二阶导数）会再乘以一个大于1的膨胀系数，再和大梯度样本的梯度（和损失函数二阶导数）进行相加，构成一个数据集的梯度（和损失函数二阶导数），来指导后续的迭代进行。而之所以要让小梯度样本进一步膨胀再加入到样本数据梯度中，其实也是为了尽可能还原原始真实的数据集梯度。也即是说，GOSS抽样并不是想要改变数据集梯度，而是希望通过更小的计算量，来尽可能还原原始数据集完整梯度，以此来提升建模的精确度，其基本过程可以由下图进行表示：<img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202303231809934.png" alt="5a9d29da9104eae4106ca796a7a493c" style="zoom:50%;" />具体膨胀系数如何计算，其实也并不复杂，就是$\frac{1-a}{b}$，或者说是$\frac{1-top\_rate}{other\_rate}$。例如当top_rate=0.1，other_rate=0.2时，小样本梯度膨胀系数为：$\frac{1-top\_rate}{other\_rate}=\frac{1-0.1}{0.2}=4.5$。最终样本梯度=大样本梯度+小样本梯度*4.5算得。

- 7.GOSS过程带来的误差：最后需要注意的是，尽管GOSS抽样后我们通过膨胀系数来尽可能还原数据集整体的梯度，但这种还原肯定是存在一定误差的。根据原论文描述，这种误差可以通过如下公式进行表示：<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/202304091801639.png" alt="1681034491832" style="zoom:50%;" /></center><br>原论文中的公式并不好理解，这里我们可以将其等价为另一个公式进行解释。首先，假设我们有一个样本集合$I$，其中有$|I|$个样本。对于一个梯度值阈值$a$，我们将样本集合$I$划分为两个子集$I_{large}$和$I_{small}$。$I_{large}$包含所有梯度值大于等于$a$的样本，而$I_{small}$包含所有梯度值小于$a$的样本。用$|I_{large}|$表示$I_{large}$中的样本数量，用$|I_{small}|$表示$I_{small}$中的样本数量。接下来，我们从$I_{small}$中随机采样一定比例（$\alpha$）的样本，则梯度计算的最大误差估计为：$$\mathrm{Error} = \frac{1}{|I|}\sum_{i \in I_{large}} g_i - \frac{1 - \alpha}{\alpha |I_{small}|}\sum_{i \in I_{small}} g_i$$这里，$g_i$表示第$i$个样本的梯度值。

#### 3.2 GOSS计算过程实例

- 损失函数、梯度与Hessian计算公式

&emsp;&emsp;接下来我们继续借助简单示例数据集data来执行一次完整的GOSS采样计算过程。由于GOSS采样计算过程会涉及样本梯度计算，因此这里我们首先需要回顾样本梯度的计算方法。我们知道样本梯度是损失函数在各参数方向上求导得到的结果，因此梯度实际上和损失函数相关。这里data数据集是二分类数据集，因此假设建模过程中的损失函数是二分类交叉熵损失函数，在Lesson 4介绍梯度下降算法时，我们就介绍了二分类交叉熵损失函数的计算公式，以及样本梯度的计算方法，并借助手写代码来进行实现，这里我们快速回顾下交叉熵损失计算公式，对于第i条样本来说，$y_i$是真实标签，而$\hat y_i$或者$p_i$是概率预测结果，则样本整体二分类交叉熵损失计算公式如下：

$$\mathcal{L}=-\frac{1}{n} \sum_{i=1}^{n}\left[y_{i} \log p_{i}+\left(1-y_{i}\right) \log \left(1-p_{i}\right)\right]
$$

此时第$i$条样本的梯度就是损失函数对预测样本的一阶偏导数：

$$\frac{\partial \mathcal{L}}{\partial p_{i}}=\frac{1-y_{i}}{1-p_{i}}-\frac{y_{i}}{p_{i}}$$

更进一步的，由于后续直方图计算过程还需要用到损失函数的二阶偏导数，也就是所谓的Hessian矩阵（值），也被称作黑塞矩阵、海森矩阵等。对于第$i$条样本，损失函数的二阶偏导数计算公式如下：

$$\frac{\partial^{2} \mathcal{L}}{\partial p_{i}^{2}}=\frac{y_{i}}{p_{i}^{2}}+\frac{1-y_{i}}{\left(1-p_{i}\right)^{2}}
$$

并且，为了方便后续计算，我们可以编写函数来进行损失函数、梯度和Hessian值的计算。其中二分类交叉熵损失可以按照如下方式进行计算：

In [3]:
def binary_cross_entropy(y_true, y_pred):
    """
    计算二分类交叉熵损失函数
    :param y_true: 真实标签，一个 Numpy 数组，形状为 (n_samples,)
    :param y_pred: 预测标签，一个 Numpy 数组，形状为 (n_samples,)
    :return:二分类交叉熵损失值，一个标量
    """
    # 确保输入的标签为 0 或 1
    y_true = np.clip(y_true, 1e-7, 1-1e-7)
    y_pred = np.clip(y_pred, 1e-7, 1-1e-7)
    # 计算二分类交叉熵损失
    loss = -(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred)).mean()
    return loss

这里的np.clip函数是用于将输出结果限制在一个范围内，避免出现零值相除的情况（sklearn中用于实现相同功能的log_loss函数也是采用类似的处理方法）。而每条样本的梯度计算函数如下：

In [4]:
def binary_cross_entropy_grad(y_true, y_pred):
    """
    计算二分类交叉熵损失函数的一阶导数
    :param y_true: 真实标签，一个 Numpy 数组，形状为 (n_samples,)
    :param y_pred: 预测标签，一个 Numpy 数组，形状为 (n_samples,)
    :return:二分类交叉熵损失函数的一阶导数，一个 Numpy 数组，形状为 (n_samples,)
    """
    # 确保输入的标签为 0 或 1
    y_true = np.clip(y_true, 1e-7, 1-1e-7)
    y_pred = np.clip(y_pred, 1e-7, 1-1e-7)
    # 计算二分类交叉熵损失函数的一阶导数
    grad = (1 - y_true) / (1 - y_pred) - y_true / y_pred
    return grad

每条样本的Hessian值计算函数如下：

In [5]:
def binary_cross_entropy_hess(y_true, y_pred):
    """
    计算二分类交叉熵损失函数的二阶导数
    :param y_true: 真实标签，一个 Numpy 数组，形状为 (n_samples,)
    :param y_pred: 预测标签，一个 Numpy 数组，形状为 (n_samples,)
    :return: 生成的特征张和标签张量：二分类交叉熵损失函数的二阶导数，一个 Numpy 数组，形状为 (n_samples,)
    """
    # 确保输入的标签为 0 或 1
    y_true = np.clip(y_true, 1e-7, 1 - 1e-7)
    y_pred = np.clip(y_pred, 1e-7, 1 - 1e-7)
    # 计算二分类交叉熵损失函数的二阶导数
    hess = y_true / (y_pred ** 2) + (1 - y_true) / ((1 - y_pred) ** 2)
    return hess

有了这些函数定义后，接下来我们来查看当前数据集在首次建模之前的梯度值如何计算。

- LGBM初始预测值

&emsp;&emsp;正如此前所说，首次进行GOSS抽样时梯度计算依据是LGBM算法的初始预测值，如果是交叉熵损失函数，则初始预测值为1类样本占比（或者1类样本总数）的对数几率（log odds）计算结果，并且每条样本初始预测值都相同，具体计算过程可以由如下计算公式表示：

$$\hat y = ln\frac{p}{1-p}$$

其中$p$为1类的占比（或者1类总数，计算结果都相同）。例如对于data数据集来说，初始预测值可以按照如下方式进行计算：

In [6]:
data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4
0  1.2  4.7   1   0  1        0.0        1.0           1.0
1  2.9  5.5   1   0  0        1.0        1.0           1.0
2  2.6  3.9   0   1  1        1.0        0.0           2.0
3  3.3  6.2   1   0  0        1.0        1.0           1.0
4  2.0  3.5   1   0  1        0.0        0.0           0.0
5  2.5  4.5   1   1  1        1.0        1.0           3.0
6  1.4  5.1   1   0  0        0.0        1.0           1.0
7  2.1  2.7   0   1  0        0.0        0.0           2.0
8  1.7  4.1   1   0  1        0.0        0.0           0.0
9  3.0  3.8   1   1  1        1.0        0.0           2.0

In [7]:
# 计算每个类别的频率
class_freq = data['y'].value_counts(normalize=True)

# 计算初始预测值
initial_prediction = np.log(class_freq[1] / class_freq[0])

print("初始预测值：", initial_prediction)

初始预测值： 0.4054651081081642


In [8]:
class_freq

1    0.6
0    0.4
Name: y, dtype: float64

In [9]:
class_freq[1]

0.6

因此所有样本的初始预测结果为0.4，将其添加到data的最后一列中：

In [10]:
y_pred = np.array([0.4054]*10)
y_pred

array([0.4054, 0.4054, 0.4054, 0.4054, 0.4054, 0.4054, 0.4054, 0.4054,
       0.4054, 0.4054])

In [11]:
data['y_pred'] = y_pred
data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4  y_pred
0  1.2  4.7   1   0  1        0.0        1.0           1.0  0.4054
1  2.9  5.5   1   0  0        1.0        1.0           1.0  0.4054
2  2.6  3.9   0   1  1        1.0        0.0           2.0  0.4054
3  3.3  6.2   1   0  0        1.0        1.0           1.0  0.4054
4  2.0  3.5   1   0  1        0.0        0.0           0.0  0.4054
5  2.5  4.5   1   1  1        1.0        1.0           3.0  0.4054
6  1.4  5.1   1   0  0        0.0        1.0           1.0  0.4054
7  2.1  2.7   0   1  0        0.0        0.0           2.0  0.4054
8  1.7  4.1   1   0  1        0.0        0.0           0.0  0.4054
9  3.0  3.8   1   1  1        1.0        0.0           2.0  0.4054

接下来我们即可依据y_pred来计算每条样本的梯度。

- 样本梯度与hes值计算

&emsp;&emsp;这里我们借助刚才定义的binary_cross_entropy_grad函数进行样本梯度计算：

In [12]:
binary_cross_entropy_grad(data['y'], data['y_pred'])

0   -2.466699
1    1.681802
2   -2.466699
3    1.681802
4   -2.466699
5   -2.466699
6    1.681802
7    1.681802
8   -2.466699
9   -2.466699
dtype: float64

并将其追加到data数据集中：

In [13]:
data['grad'] = binary_cross_entropy_grad(data['y'], data['y_pred'])
data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4  y_pred      grad
0  1.2  4.7   1   0  1        0.0        1.0           1.0  0.4054 -2.466699
1  2.9  5.5   1   0  0        1.0        1.0           1.0  0.4054  1.681802
2  2.6  3.9   0   1  1        1.0        0.0           2.0  0.4054 -2.466699
3  3.3  6.2   1   0  0        1.0        1.0           1.0  0.4054  1.681802
4  2.0  3.5   1   0  1        0.0        0.0           0.0  0.4054 -2.466699
5  2.5  4.5   1   1  1        1.0        1.0           3.0  0.4054 -2.466699
6  1.4  5.1   1   0  0        0.0        1.0           1.0  0.4054  1.681802
7  2.1  2.7   0   1  0        0.0        0.0           2.0  0.4054  1.681802
8  1.7  4.1   1   0  1        0.0        0.0           0.0  0.4054 -2.466699
9  3.0  3.8   1   1  1        1.0        0.0           2.0  0.4054 -2.466699

然后再计算样本的hes值：

In [14]:
binary_cross_entropy_hess(data['y'], data['y_pred'])

0    6.084606
1    2.828461
2    6.084606
3    2.828461
4    6.084606
5    6.084606
6    2.828461
7    2.828461
8    6.084606
9    6.084606
dtype: float64

In [15]:
data['hess'] = binary_cross_entropy_hess(data['y'], data['y_pred'])
data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4  y_pred      grad  \
0  1.2  4.7   1   0  1        0.0        1.0           1.0  0.4054 -2.466699   
1  2.9  5.5   1   0  0        1.0        1.0           1.0  0.4054  1.681802   
2  2.6  3.9   0   1  1        1.0        0.0           2.0  0.4054 -2.466699   
3  3.3  6.2   1   0  0        1.0        1.0           1.0  0.4054  1.681802   
4  2.0  3.5   1   0  1        0.0        0.0           0.0  0.4054 -2.466699   
5  2.5  4.5   1   1  1        1.0        1.0           3.0  0.4054 -2.466699   
6  1.4  5.1   1   0  0        0.0        1.0           1.0  0.4054  1.681802   
7  2.1  2.7   0   1  0        0.0        0.0           2.0  0.4054  1.681802   
8  1.7  4.1   1   0  1        0.0        0.0           0.0  0.4054 -2.466699   
9  3.0  3.8   1   1  1        1.0        0.0           2.0  0.4054 -2.466699   

       hess  
0  6.084606  
1  2.828461  
2  6.084606  
3  2.828461  
4  6.084606  
5  6.084606  
6  2.828461  
7  2.828461  
8  6.084606  
9  6.084606

- GOSS抽样

&emsp;&emsp;据此我们即可进一步执行GOSS抽样。这里我们假设top_rate=0.2，other_rate=0.5，由于我们样本量较少，因此这里设置的比例偏大。此时GOSS抽样过程如下，首先计算样本梯度绝对值：

In [16]:
abs_gradients = np.abs(data['grad'])
abs_gradients

0    2.466699
1    1.681802
2    2.466699
3    1.681802
4    2.466699
5    2.466699
6    1.681802
7    1.681802
8    2.466699
9    2.466699
Name: grad, dtype: float64

然后进行从大到小的索引排序：

In [17]:
sorted_index = np.argsort(-abs_gradients)
sorted_index

0    0
1    2
2    4
3    5
4    8
5    9
6    1
7    3
8    6
9    7
Name: grad, dtype: int64

此时top_rate=0.2，other_rate=0.5，即从10条数据中挑选出梯度最大的2条数据，然后从剩下的8条数据中抽取50%，即抽取4条数据，共同构成本次GOSS抽样得到的训练数据集。这里我们先借助布尔索引先挑选出梯度最大的2条数据：

In [18]:
topn_index = sorted_index[:2]
topn_index.values

array([0, 2])

两条数据抽取结果如下：

In [19]:
topn_data = data.iloc[topn_index.values, :]
topn_data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4  y_pred      grad  \
0  1.2  4.7   1   0  1        0.0        1.0           1.0  0.4054 -2.466699   
2  2.6  3.9   0   1  1        1.0        0.0           2.0  0.4054 -2.466699   

       hess  
0  6.084606  
2  6.084606

然后再从剩下的数据集中随机抽取4条数据，首先是找到剩余数据集索引：

In [20]:
sorted_index[2:]

2    4
3    5
4    8
5    9
6    1
7    3
8    6
9    7
Name: grad, dtype: int64

然后进行抽样，这里采用无放回抽样：

In [57]:
raten_index = np.random.choice(sorted_index[2:].values, size=4, replace=False)
raten_index

array([5, 8, 1, 3], dtype=int64)

In [58]:
raten_data = data.iloc[raten_index, :]
raten_data

x1   x2  x3  x4  y  x1_binned  x2_binned  x2_binned&x4  y_pred      grad  \
5  2.5  4.5   1   1  1        1.0        1.0           3.0  0.4054 -2.466699   
8  1.7  4.1   1   0  1        0.0        0.0           0.0  0.4054 -2.466699   
1  2.9  5.5   1   0  0        1.0        1.0           1.0  0.4054  1.681802   
3  3.3  6.2   1   0  0        1.0        1.0           1.0  0.4054  1.681802   

       hess  
5  6.084606  
8  6.084606  
1  2.828461  
3  2.828461

至此，我们就完成了一次GOSS抽样全过程，接下来我们就将带入我们抽样得到的topn_data和raten_data带入进行模型训练。

---

- 课程节选自《机器学习实战训练营》正式课程

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/1.jpg" alt="1" style="zoom:20%;" />

&emsp;&emsp;公开课内容节选自《机器学习实战训练营》正式课程，正式课程相比公开课，额外增加更具深度和LGBM数学原理推导、LGBM原生API使用和优化技巧，以及基于LGBM原生API的企业级实战案例，公开课与付费正课内容对比如下：

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/05b61eb9b6792c10ce8c6c221f7d140.png" alt="05b61eb9b6792c10ce8c6c221f7d140" style="zoom:50%;" />

&emsp;&emsp;如果同学希望更进一步进行深度、系统的学习LightGBM及其他机器学习算法，欢迎大家报名《机器学习实战训练营（第六期）》付费正课，课程为130+小时完整体系大课，完整涵盖经典机器学习、集成学习、特征工程、模型融合、特征混合增强和企业级之战案例，总共6大模块，零基础直达机器学习算法岗中高级岗位要求，六年教学经验沉淀，千名学员口碑见证，超值体系大课等你来学！

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230215185442588.png" alt="image-20230215185442588" style="zoom:40%;" />

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230215185629296.png" alt="image-20230215185629296" style="zoom:42.5%;" />

&emsp;&emsp;除了《机器学习实战训练营》课程外，我们还开设了《数据分析就业班》和《深度学习实战课》，在此查看详细课程介绍和课程大纲：https://appZe9inzwc2314.h5.xiaoeknow.com：

<center><img src="http://ml2022.oss-cn-hangzhou.aliyuncs.com/img/image-20230214180459436.png" alt="image-20230214180459436" style="zoom:50%;" />

- **<font color='red'>第六期训练营课程尾单八折特惠进行时！**

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/2.jpg" alt="2" style="zoom:30%;" />

&emsp;&emsp;<font size=3><center>扫码添加客服“小可爱”，或者手动添加客服微信：littlecat_1205       
    >>>  <font color='red'>**回复“优惠”**</font>抢<font color='red'>**直播间限定折上折**        
    **>>>  限量赠送授课老师亲自答疑服务！**

<center><img src="https://ml2022.oss-cn-hangzhou.aliyuncs.com/img/926380675d5976b65f5784be6db5f35.png" alt="926380675d5976b65f5784be6db5f35" style="zoom:50%;" />